### DISCUSSION TITLE
kv.v2.list_secrets fails when LIST request not supported

### DISCUSSION THREAD

**Discussion by RomanTheLegend** [hvac/hvac#947] [Q&A]
I've stumbled upon a problem that when `LIST` request is not supported, the library cannot extract the list of secrets from Vault, while the default client can (because it uses `GET`)

`kv.v2.list_secrets(path="mysecret")` will throw "Bad Request" exception

The problem is in this block:
https://github.com/hvac/hvac/blob/develop/hvac/api/secrets_engines/kv_v2.py#L328

Changing it to GET fixes the issue:

```
        api_path = utils.format_url(
            "/v1/{mount_point}/metadata/{path}?list=true", mount_point=mount_point, path=path
        )
        return self._adapter.get(
            url=api_path,
        )
```

for the reference, the official client uses GET too:

```
vault kv list -output-curl-string  secret/mysecret 
curl -H "X-Vault-Request: true" -H "X-Vault-Token: $(vault print token)" https://vault.example.com/v1/secret/metadata/mysecret?list=true
```

It would be good to have an overloaded method that explicitly uses GET as a request

**Comment by namenmalkav** [comment]
im having the same issue

**Comment by RomanTheLegend** [comment]
> im having the same issue

Here's my quick&ugly workaround until the lib is fixed:

```
import json
import os
import requests


def list_secrets(folder):
    url = "{}/v1/secret/metadata/{}".format(os.environ['VAULT_ADDR'],requests.utils.quote(folder,safe=''))
    print(url)

    headers = {"X-Vault-Token" : os.environ['VAULT_TOKEN']}
    res = requests.get(url, params={"list":"true"} , headers=headers).content
    children = json.loads(res)["data"]["keys"]
    print(children)
    return children
```

**Comment by briantist** [comment]
Hi folks! We do have a method to force the use of `GET` requests for `LIST` operations. 
To do so, you should create your own Adapter type. This could be as easy as subclassing the default [`JSONAdapter`](https://hvac.readthedocs.io/en/stable/source/hvac_adapters.html#hvac.adapters.JSONAdapter):

```python
import hvac

class StrictAdapter(hvac.adapters.JSONAdapter):
    def __init__(self, *args, **kwargs):
        kwargs["strict_http"] = True
        super().__init__(*args, **kwargs)

client = hvac.Client(adapter=StrictAdapter)
```

---

To follow the flow of how that's used:

In the file linked in the OP, we see it the method call `adapter.list`:
https://github.com/hvac/hvac/blob/d7e9557c29d9d4f42d2c8ebe84bccbc26e505873/hvac/api/secrets_engines/kv_v2.py#L330

`adapter.list` calls `adapter.request` with the `"list"` method:
https://github.com/hvac/hvac/blob/d7e9557c29d9d4f42d2c8ebe84bccbc26e505873/hvac/adapters.py#L162

which in turn has code to check whether the method is `list` and if so, to change the request method to `GET` and add the `list=true` parameter, but only when `strict_http` is set to `True` (it defaults to `False`):
https://github.com/hvac/hvac/blob/d7e9557c29d9d4f42d2c8ebe84bccbc26e505873/hvac/adapters.py#L295-L301

---

I think there's room for improvement here though.

An overload, or a new parameter is not quite what I'd recommend, because all list methods would need the same thing, not just kv2, and the issue of needing to use `GET` instead of `LIST` should be system wide (which is why it's in the adapter).

Instead I'd probably look to make one or more of the following changes:
- Allow the client to take a pre-created instance of an Adapter, and not just the class: it would be much easier to simply pass it in than to have to subclass it, `hvac.Client(adapter=hvac.adapters.JSONAdapter(strict_http=True))`
- Perhaps add a `strict_http` option to the `Client` constructor so it can be set there, to be passed into the Adapter, `hvac.Client(strict_http=True)`

I'll think about it a bit and open an issue, and probably a PR.

**Reply by jtele2** [reply]
Did this ever get solved? I'm still seeing the issue in hvac 2.3.0.

**Reply by alexcfpho** [reply]
Having the same issue in `2.4.0` now, works when setting `strict_http=True`.

**Reply by Tylerlhess** [reply]
yeah the check is dumb in my opinion and misses if it isn't strict_http

https://github.com/hvac/hvac/blob/097bfdebfd354db7b1913c4a3384eeef6c41f0fd/hvac/adapters.py#L323-L330

is where the issue is. Not sure why "strict_http" matters to be a pre-req for this code block.

**Reply by Tylerlhess** [reply]
submitted a pr for it.
